server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java

```diff
@@ -0,0 +1,361 @@
+            // filter is full, create a new one and insert there
+            CuckooFilter t = new CuckooFilter(capacity, fpp, rng);
+            t.add(hash);
+            filters.add(t);
+            breaker.accept(t.getSizeInBytes()); // make sure we account for the new filter
+        }
+        return true;
+    }
+
+    /**
+     * If we still holding values in a set, convert this filter into an approximate, cuckoo-backed filter.
+     * This will create a list of CuckooFilters, and null out the set of hashes
+     */
+    private void convert() {
+        if (isSetMode) {
+            long oldSize = getSizeInBytes();
+
+            filters = new ArrayList<>();
+            CuckooFilter t = new CuckooFilter(capacity, fpp, rng);
+            hashes.forEach(t::add);
+            filters.add(t);
+
+            hashes = null;
+            isSetMode = false;
+
+            breaker.accept(-oldSize); // this zeros out the overhead of the set
+            breaker.accept(getSizeInBytes()); // this adds back in the new overhead of the cuckoo filters
+        }
+    }
+
+    /**
+     * Get the approximate size of this datastructure.  Approximate because only the Set occupants
+     * are tracked, not the overhead of the Set itself.
+     */
+    public long getSizeInBytes() {
+        long bytes = 0;
+        if (hashes != null) {
+            bytes = (hashes.size() * 16) + 8 + 4 + 1;
+        }
+        if (filters != null) {
+            bytes += filters.stream().mapToLong(CuckooFilter::getSizeInBytes).sum();
+        }
+        return bytes;
+    }
+
+
+    /**
+     * Merge `other` cuckoo filter into this cuckoo.  After merging, this filter's state will
+     * be the union of the two.  During the merging process, the internal Set may be upgraded
+     * to a cuckoo if it goes over threshold
+     */
+    public void merge(SetBackedScalingCuckooFilter other) {
+        if (isSetMode && other.isSetMode) {
+            // Both in sets, merge collections then see if we need to convert to cuckoo
+            hashes.addAll(other.hashes);
+            if (hashes.size() > threshold) {
+                convert();
+            }
+        } else if (isSetMode && other.isSetMode == false) {
+            // Other is in cuckoo mode, so we convert our set to a cuckoo then merge collections.
+            // We could probably get fancy and keep our side in set-mode, but simpler to just convert
+            convert();
+            filters.addAll(other.filters);
+        } else if (isSetMode == false && other.isSetMode) {
+            // Rather than converting the other to a cuckoo first, we can just
+            // replay the values directly into our filter.
+            other.hashes.forEach(this::add);
+        } else {
+            // Both are in cuckoo mode, merge raw fingerprints
+
+            int current = 0;
+            CuckooFilter currentFilter = filters.get(current);
+
+            for (CuckooFilter filter : other.filters) {
```

Comment:
nit: for clarity could we call the variable `otherFilter`? This might avoid confusion below on whether its the filter in this instance or the `other` instance